### DISCUSSION TITLE
How is the moving average calculated?

### DISCUSSION THREAD

**Discussion by jnoche** [mortbopet/Ripes#251] [Q&A] [answered]
I have a 7-word program and the (32-entry 4-word direct-mapped) instruction cache for it gets "miss hit hit hit miss hit" for the first 6 cycles.  The graph in Ripes for the hit rates (together with a table I made) is shown below:

<img width="420" alt="temp" src="https://user-images.githubusercontent.com/63948655/204119476-9f9033b7-65b4-450e-b96b-b4ed74490623.png">

The calculation for the cumulative hit rate is straightforward, but I've spent a few days on it and I still cannot understand how the (50-cycle) moving average is calculated even after I looked at the source code.

```
    // Moving average plot
    if (m_ui->windowed->isChecked()) {
      QPointF dNum, dDen;
      if (!m_lastDiffValid) {
        dNum = p1;
        dDen = p2;
        m_lastDiffValid = true;
      } else {
        dNum = p1 - m_lastDiffData.first;
        dDen = p2 - m_lastDiffData.second;
      }
      double wRatio = 0;
      if (dDen.y() != 0) {
        wRatio = static_cast<double>(dNum.y()) / dDen.y();
        wRatio *= 100.0;
      }

      m_maxY = wRatio > m_maxY ? wRatio : m_maxY;
      m_minY = wRatio < m_minY ? wRatio : m_minY;
      m_mavgData.push(wRatio);

      // Plot moving average
      const double wAvg =
          std::accumulate(m_mavgData.begin(), m_mavgData.end(), 0.0) /
          m_mavgData.size();
      newWindowPoints << QPointF(p1.x(), wAvg);
      m_lastDiffData.first = p1;
      m_lastDiffData.second = p2;
    }
```

Could you show me how the values in the last row of my table above were calculated?

**Comment by mortbopet** [comment, marked as answer]
Copy-pasting from [211a420](https://github.com/mortbopet/Ripes/commit/211a420c249c55561cf17e9c3db21e4dc78906f7)

Moving average was calculated incorrectly, as shown in #251. I can't recall the exact reason for the old implementation, but my hunch is that i wanted the moving average to be able to be plotted in the "total hits/total access" configuration, which led to this notion of tracking last iteration differences etc. etc..
However, this methodology is flawed to begin with, since the moving average (obviously, in hindsight) need to be calculated on a a range of "was cycle # a hit or miss" values across a given window.

A moving average can now be achieved by setting "was hit / 1" as the configuration, and enabling the moving average (and disabling the "ratio" plot). I still think allowing a user to freely mix and max numerators and denominators to their liking is a good design descision, however, i could also easily imagine that some people wouldn't immediatly intuit themselves into the exact configuration needed for a moving average to show.
Secondly, i'd like to have the option of superimposing multiple plots - this would be needed to see e.g. the relation between moving average hitrate and total hitrate.

Both of these features need a bit more work to be implemented correctly, so that will have to be in a separate, followup commit.
